back/engines/commercial/project_management/app/models/project_management/patches/user.rb

```diff
@@ -24,17 +25,54 @@ def enabled_roles
         end
       end
 
-      def project_moderator?(project_id = nil)
-        project_id ? moderatable_project_ids.include?(project_id) : moderatable_project_ids.present?
-      end
-
       def moderatable_project_ids
         roles.select { |role| role['type'] == 'project_moderator' }.pluck("project_id").compact
       end
 
+      def moderatable_project_ids_was
+        roles_was.select { |role| role['type'] == 'project_moderator' }.pluck("project_id").compact
+      end
+
       def moderatable_projects
         ::Project.where(id: moderatable_project_ids)
       end
+
+      def moderatable_projects_was
+        ::Project.where(id: moderatable_project_ids_was)
+      end
+
+      def project_moderator?(project_id = nil)
+        project_id ? moderatable_project_ids.include?(project_id) : moderatable_project_ids.present?
+      end
+
+      module IdeaAssignment
+        def self.included(base)
+          base.before_save :reset_project_and_idea_assignments, if: :lost_rights_over_assigned_projects_or_ideas?
+        end
+
+        # When a user loses project management rights over a project, and is the default assignee, or assignee of some ideas,
+        # change it to the oldest admin.
+        def reset_project_and_idea_assignments
+          projects = lost_assigned_moderatable_projects
+          ideas = lost_assigned_moderated_ideas
+          projects.update(default_assignee: self.class.oldest_admin)
+          ideas.update(assignee: self.class.oldest_admin)
+        end
+
+        def lost_rights_over_assigned_projects_or_ideas?
+          return false if admin?
+
+          lost_assigned_moderatable_projects.any? || lost_assigned_moderated_ideas.any?
+        end
+
+        def lost_assigned_moderatable_projects
+          ::Project.where(id: moderatable_project_ids_was - moderatable_project_ids, default_assignee: self)
+        end
+
+        def lost_assigned_moderated_ideas
+          Idea.includes(:project).where(project: ::Project.where(id: moderatable_project_ids_was - moderatable_project_ids), assignee: self)
+        end
+      end
```

Comment:
This logic lives in the intersection between idea assignment and project management modules, so I was a bit confused as where to place it. Decided to go for this, a module called `IdeaAssignment` contained in the `ProjectManagement` user patch. This should be enough to understand what it is related to